File: dotnet/src/MableBank.Application/LoadAccountBalances.cs
using MableBank.Domain;

namespace MableBank.Application;

public static class LoadAccountBalances
{
    public static async Task<CompanyAccountBook> FromReaderAsync(
        IAccountBalanceReader reader,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(reader);
        var book = new CompanyAccountBook();
        await foreach (var account in reader.ReadAccountsAsync(cancellationToken).ConfigureAwait(false))
        {
            book.AddAccount(account);
        }

        book.Ledger = new RollingBackLedger(book);
        return book;
    }
}

File: ruby/lib/mable_bank/application/load_account_balances.rb
# frozen_string_literal: true

module MableBank
  module Application
    class LoadAccountBalances
      def initialize(reader:)
        @reader = reader
      end

      def call
        book = Domain::CompanyAccountBook.new
        stream_accounts { |account| book.add_account(account) }
        book.ledger = RollingBackLedger.new(book)
        book
      end

      private

      def stream_accounts(&)
        if @reader.respond_to?(:each_account)
          @reader.each_account(&)
        else
          @reader.read.each(&)
        end
      end
    end
  end
end
